micromamba/src/update.cpp

```diff
@@ -8,10 +8,121 @@
 
 #include "mamba/api/configuration.hpp"
 #include "mamba/api/update.hpp"
+#include "mamba/api/channel_loader.hpp"
 
+#include "mamba/core/transaction.hpp"
+#include "mamba/core/context.hpp"
+#include "mamba/core/util_os.hpp"
+#include "version.hpp"
+
+extern "C"
+{
+#include "solv/evr.h"
+#include "solv/pool.h"
+#include "solv/conda.h"
+#include "solv/repo.h"
+#include "solv/selection.h"
+#include "solv/solver.h"
+}
 
 using namespace mamba;  // NOLINT(build/namespaces)
 
+int
+update_self(const std::optional<std::string>& version)
+{
+    auto& config = mamba::Configuration::instance();
+    auto& ctx = mamba::Context::instance();
+    config.load();
+
+    mamba::MPool pool;
+    mamba::MultiPackageCache package_caches(ctx.pkgs_dirs);
+
+    auto exp_loaded = load_channels(pool, package_caches, 0);
+    if (!exp_loaded)
+    {
+        throw std::runtime_error(exp_loaded.error().what());
+    }
+
+    pool.create_whatprovides();
+    std::string matchspec;
+    if (!version)
+    {
+        matchspec = fmt::format("micromamba>{}", umamba::version());
+    }
+    else
+    {
+        matchspec = fmt::format("micromamba={}", version.value());
+    }
+
+    auto solvable_ids = pool.select_solvables(pool.matchspec2id(matchspec), true);
+
+    if (solvable_ids.empty())
+    {
+        if (pool.select_solvables(pool.matchspec2id("micromamba")).empty())
+        {
+            throw std::runtime_error(
+                "micromamba not found in the loaded channels. Add 'conda-forge' to your config file.");
+        }
+        else
+        {
+            Console::instance().print(fmt::format(
+                "\nYour micromamba version ({}) is already up to date.", umamba::version()));
+            return 0;
+        }
+    }
+
+    std::optional<PackageInfo> latest_micromamba = pool.id2pkginfo(solvable_ids[0]);
+    LOG_WARNING << latest_micromamba.value().url;
+
+    ctx.download_only = true;
+    MTransaction t(pool, { latest_micromamba.value() }, package_caches);
+    auto exp_prefix_data = PrefixData::create(ctx.root_prefix);
+    if (!exp_prefix_data)
+    {
+        // TODO: propagate tl::expected mechanism
+        throw std::runtime_error(exp_prefix_data.error().what());
+    }
+    PrefixData& prefix_data = exp_prefix_data.value();
+    t.execute(prefix_data);
+
+    fs::u8path mamba_exe = get_self_exe_path();
+    fs::u8path mamba_exe_bkup = mamba_exe;
+    mamba_exe_bkup.replace_extension(".bkup");
+
+    fs::u8path cache_path = package_caches.get_extracted_dir_path(latest_micromamba.value())
+                            / latest_micromamba.value().str();
+
+    fs::rename(mamba_exe, mamba_exe_bkup);
+
+    try
+    {
+        if (on_win)
+        {
+            fs::copy_file(cache_path / "Library" / "bin" / "micromamba.exe",
```

Comment:
Not sure if this is completely relevant to this specific PR, but on my windows install, in my environments, I cannot find micromamba in `Library/bin`. Is my setup broken or is it a possibility? If it is, should that be handled here?

Reply:
If you check the `micromamba` package from conda forge you should find the binary file there though. It's the 'standard' layout for conda packages on Windows.